### PR TITLE
[corlib] corefx tests from System.Threading.* and System.Text.RegularExpressions

### PR DESCRIPTION
**Pull request by EgorBo**
Tests from the following corefx assemblies:

- System.Threading.Tasks
- System.Threading.Tasks.Parallel
- System.Threading.ThreadPool
- System.Threading.Timer
- System.Threading.Thread
- System.Text.RegularExpressions
- System.Threading

Total: 1624, Errors: 0, Failed: 49, Skipped: 2, Time: 127.351s

### CODE REVIEW

File: mcs/class/test-helpers/PlatformDetection.cs [hunk 1]
```diff
@@ -3,5 +3,10 @@ namespace System
 	static partial class PlatformDetection
 	{
 		public static readonly bool IsNetNative = false;
+		public static readonly bool IsNotWinRT = true;
+		public static readonly bool IsWinRT = false;
+		public static readonly bool IsFullFramework = true;
+		public static readonly bool IsWindowsNanoServer = false;
+		public static readonly bool IsNonZeroLowerBoundArraySupported = false;
```
Comment (marek-safar) [inline mcs/class/test-helpers/PlatformDetection.cs]: I think this one could be true

File: mcs/class/test-helpers/PlatformDetection.cs [hunk 2]
```diff
@@ -3,9 +3,10 @@ namespace System
 	static partial class PlatformDetection
 	{
 		public static readonly bool IsNetNative = false;
-
+		public static readonly bool IsNotWinRT = true;
+		public static readonly bool IsWinRT = false;
 		public static readonly bool IsFullFramework = true;
-
-		public static bool IsNonZeroLowerBoundArraySupported => true;
```
Comment (marek-safar) [inline mcs/class/test-helpers/PlatformDetection.cs]: IsNonZeroLowerBoundArraySupported has to be a property, corefx testing framework depends on that

File: mcs/build/gensources.sh [hunk 3]
```diff
@@ -1,4 +1,4 @@
-#! /bin/sh
+#!/usr/bin/env bash
```
Comment (marek-safar) [inline mcs/build/gensources.sh]: Why is this needed

Comment (EgorBo) [reply mcs/build/gensources.sh]: I use some bash functions like [[ - will fail on sh. other .sh files also use it https://github.com/mono/mono/blob/master/mcs/tools/removecomments.sh

Comment (marek-safar) [reply mcs/build/gensources.sh]: right but this way it'll look up for anything named `bash` not necessary bash shell

**Comment by marek-safar** [comment]
build

File: mcs/build/gensources.sh (hunk 3 above)
Comment (akoeplinger) [reply mcs/build/gensources.sh]: some Unix systems don't have bash in /bin so using /usr/bin/env is usually considered more portable.

I do wonder though whether we shouldn't stick to /bin/sh instead and removing the bashisms for such a core script (afaik all the other scripts that use bash aren't required for the build).

Comment (marek-safar) [reply mcs/build/gensources.sh]: @akoeplinger  I don't agree this is better, it means that the script could behave differently depending on who runs it.

Comment (EgorBo) [reply mcs/build/gensources.sh]: @marek-safar @akoeplinger btw, all .net core build scripts also use "/usr/bin/env bash". I actually changed sh to bash because of linux build machine - macOS and windows(cygwin) work with sh as a first line just fine (as well as my local linux machine). I'll try to rewrite it to sh if it's possible.
